src/v/cluster/tm_stm_types.cc

```diff
@@ -10,24 +10,99 @@
  */
 #include "cluster/tm_stm_types.h"
 
+#include <optional>
+
 namespace cluster {
 
+namespace {
+template<typename... Args>
+bool is_one_of(tx_status status, Args&&... args) {
+    return ((args == status) || ...);
+}
+
+} // namespace
+
+state_transition_error::state_transition_error(tx_status from, tx_status to)
+  : from(from)
+  , to(to) {}
+
+bool is_state_transition_valid(
+  const tx_metadata& current, tx_status requested_status) {
+    /**
+     * The state transition validation logic is really a simple definition of
+     * transaction fsm. The validation checks it the current state is a valid
+     * precursor of the requested one.
+     */
+    switch (requested_status) {
+    case empty:
+        // ready is an initial state a transaction can never go back to that
+        // state
+        return is_one_of(current.status, tx_status::empty);
```

Comment:
nit: I think one confusing bit here is the state transition from to itself (current == target) is a valid transition. Is that intentional? That seems to be the case for all the states.

Reply:
yes, this is intentional for the transitions to be idempotent, i will think more about this